improver/calibration/ensemble_calibration.py

```diff
@@ -134,17 +142,237 @@ def __init__(self, tolerance=0.02, max_iterations=1000):
+
+        Return:
+            numpy.ndarray:
+                A single set of coefficients with the order [alpha, beta, gamma, delta].
+
+        """
+        optimised_coeffs = minimize(
+            minimisation_function,
+            initial_guess,
+            args=(
+                forecast_predictor_data,
+                truth_data,
+                forecast_var_data,
+                sqrt_pi,
+                predictor,
+            ),
+            method="Nelder-Mead",
+            tol=self.tolerance,
+            options={"maxiter": self.max_iterations, "return_all": True},
+        )
+
+        return optimised_coeffs
+
+    def _process_points_independently(
+        self,
+        minimisation_function,
+        initial_guess,
+        forecast_predictor,
+        truth,
+        forecast_var,
+        predictor,
+    ):
+        """Minimise each point along the spatial dimensions independently to
+        create a set of coefficients for each point. The coefficients returned
+        can be either gridded (i.e. separate dimensions for x and y) or for a
+        list of sites where x and y share a common dimension.
+
+        Args:
+            minimisation_function (method instance):
+                Function to use when minimising.
+            initial_guess (list)
+            forecast_predictor (iris.cube.Cube)
+            truth (iris.cube.Cube)
+            forecast_var (iris.cube.Cube)
+            predictor (str)
+
+        Returns:
+            numpy.ndarray:
+                Separate optimised coefficients for each point. The shape of the
+                coefficients array is (number of coefficients, length of spatial dimensions).
+                Order of coefficients is [alpha, beta, gamma, delta].
+        """
+        initial_guess = np.array(initial_guess, dtype=np.float64)
+        forecast_predictor.data = forecast_predictor.data.astype(np.float64)
+        forecast_var.data = forecast_var.data.astype(np.float64)
+        truth.data = truth.data.astype(np.float64)
+        sqrt_pi = np.sqrt(np.pi).astype(np.float64)
+
+        argument_list = []
+        sindex = [
+            forecast_predictor.coord(axis="y"),
+            forecast_predictor.coord(axis="x"),
+        ]
+
+        for index, (fp_slice, truth_slice, fv_slice) in enumerate(
+            zip(
+                forecast_predictor.slices_over(sindex),
+                truth.slices_over(sindex),
+                forecast_var.slices_over(sindex),
+            )
+        ):
+            argument_list.append(
+                (
+                    minimisation_function,
+                    initial_guess[index],
+                    fp_slice.data.T,
+                    truth_slice.data,
+                    fv_slice.data,
+                    sqrt_pi,
+                    predictor,
+                )
+            )
+
+        with Pool(os.cpu_count()) as pool:
+            optimised_coeffs = pool.starmap(self._minimise_caller, argument_list)
```

Comment:
I've tried using the `starmap_async` option, however, there was no benefit in terms of speed. I think this is because `starmap` and `starmap_async` only differ in terms of blocking, so `starmap_async` could have benefit if there was some code in between setting off the parallel processing and getting the results from the parallel processing. In this code, there isn't anything in between, so that's why I'm don't get any benefit from `starmap_async`. There's a bit more information about this on this [StackOverflow answer](https://stackoverflow.com/questions/35908987/multiprocessing-map-vs-map-async?answertab=votes#tab-top).

For the record, I've also compared this with a few approaches using Dask: https://examples.dask.org/applications/embarrassingly-parallel.html. I tried the `delayed` and `concurrent.Futures` approaches, as a replacement for the multiprocessing but they didn't give any benefit, in terms of speed.

Using `delayed`:
```
        client = Client(n_workers=os.cpu_count())
        results = []
        for index, (fp_slice, truth_slice, fv_slice) in enumerate(
            zip(
                forecast_predictor.slices_over(sindex),
                truth.slices_over(sindex),
                forecast_var.slices_over(sindex),
            )
        ):
            coeffs = delayed(self._minimise_caller)(minimisation_function,
                    initial_guess[index],
                    fp_slice.data.T,
                    truth_slice.data,
                    fv_slice.data,
                    sqrt_pi,
                    predictor)
            results.append(coeffs)
        optimised_coeffs = dask.compute(*results, scheduler="processes")
        client.close()
```
Using `concurrent.Futures` (this is for the initial guess processing, but the idea is the same):
```
            client = Client(n_workers=os.cpu_count())
            futures = []
            for (truths_slice, fp_slice) in zip(
                truths.slices_over(index), forecast_predictor.slices_over(index)
            ):
                future = client.submit(self.compute_initial_guess, truths_slice.data,
                    fp_slice.data,
                    self.predictor,
                    False
                    if np.any(np.isnan(truths_slice.data))
                    or np.any(np.isnan(fp_slice.data))
                    else self.ESTIMATE_COEFFICIENTS_FROM_LINEAR_MODEL_FLAG,
                    number_of_realizations,
                )
                futures.append(future)

            initial_guess = client.gather(futures)
            client.close()
```

Reply:
Thanks for looking into it.